### PR TITLE
hackagePackages: dontCheck cryptohash-sha512, dontCheck hnix-store-remote

### PR DESCRIPTION
**Pull request by Anton-Latukha**
*Status: waiting on the https://github.com/commercialhaskell/stackage/issues/5766 & https://github.com/haskell-nix/hnix-store/issues/104*

###### Motivation for this change

Bringing up to date configuration of a couple of packages.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 1]
```diff
@@ -816,7 +819,7 @@ self: super: {
   # See https://github.com/haskell-hvr/cryptohash-sha512/pull/3
   # , https://github.com/haskell-hvr/cryptohash-sha512/issues/4
   # and https://github.com/haskell-hvr/cryptohash-sha512/pull/5
-  cryptohash-sha512 = doJailbreak super.cryptohash-sha512;
+  cryptohash-sha512 = dontCheck (doJailbreak super.cryptohash-sha512);
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix:822]: Why did you add `dontCheck` here?  `cryptohash-sha512` appears to currently be building even with the tests enabled.

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-common.nix:822]: Oh, I see the explanation is in the commit.

Could you add this explanation in a comment above this line?  Also, ideally link to an issue upstream where the nix maintainers can track the status of this.  (Although it seems like cryptohash-sha512 has been somewhat abandoned?)

> base16-bytestring had new epoch (1.0) 3 month ago which fixed the old return type to normal `Either String ByteString`. `cryptohash-sha512` is stale for ~10-11 months and testsuite does not support the new `base16-bytesting` releases.

Ah, I guess this new version of base16-bytestring hasn't made its way to the Stackage nightlies yet, which is why cryptohash-sha512 doesn't yet need `dontCheck`: https://github.com/commercialhaskell/stackage/issues/5649

Comment (Anton-Latukha) [reply pkgs/development/haskell-modules/configuration-common.nix:831]: Yes. It is a still working remnant of an old story.

Since it is a well-known public story I would refresh: 
When `vincenthz` was preamptively deprecating his old working `cryptohash` in favor of the new not fully formed `cryptonite`, somewhere in that time they _talked about something_. `hvr` forked the `cryptohash` and split the project into `cryptohash-*` family of more lightweight packages. In parallel, there was a debacle with the `snoyman`. And with everything, all those went to public boards and records.

No reason to say with the number of things `hvr` has on the plate. The whole `cryptohash-*` family pretty much got abandoned right after and these years rolled on merging of PRs from `phadej` & drive-by people. But it seems `hvr` stopped appearing in the https://github.com/haskell-hvr ~9-10 months ago, IDK his patterns, maybe he goes in the once a year. But probably his participation understandably dropped.

 It seems to many of those projects only he has merge rights, `cryptohash-*` family is most probably RIP as a bunch of other projects there, so you probably can await others from that group.

**Comment by cdepillabout** [comment]
Hmm, I tried building `hnix-store-remote`, but it doesn't look like it is able to build:

```console
$ git rev-parse HEAD
2ad09b6771bc76f5691cbd612f847528daceff9a
$ NIXPKGS_ALLOW_BROKEN=1 nix-build -A haskellPackages.hnix-store-remote
these derivations will be built:
  /nix/store/fwg6v2p9f9mnrjvq6836v40pwbi69ai1-hnix-store-remote-0.3.0.0.drv
building '/nix/store/fwg6v2p9f9mnrjvq6836v40pwbi69ai1-hnix-store-remote-0.3.0.0.drv'...
setupCompilerEnvironmentPhase
Build with /nix/store/kfc4428g15gzv5xq4wi7vjs1m2qzkkqh-ghc-8.10.3.
unpacking sources
unpacking source archive /nix/store/vpkdq310w3bz1h3rml8106b1qanv2s9r-hnix-store-remote-0.3.0.0.tar.gz
source root is hnix-store-remote-0.3.0.0
setting SOURCE_DATE_EPOCH to timestamp 1000000000 of file hnix-store-remote-0.3.0.0/tests/Util.hs
patching sources
compileBuildDriverPhase
setupCompileFlags: -package-db=/build/setup-package.conf.d -j8 +RTS -A64M -RTS -threaded -rtsopts
[1 of 1] Compiling Main             ( /nix/store/4mdp8nhyfddh7bllbi7xszz7k9955n79-Setup.hs, /build/Main.o )
Linking Setup ...
configuring
configureFlags: --verbose --prefix=/nix/store/0mzmlb7sx5ynlx76ylfcw1l6k3gz6a6f-hnix-store-remote-0.3.0.0 --libdir=$prefix/lib/$compiler --libsubdir=$abi/$libname --docdir=/nix/store/q97drw4qk265478rhjkv5kk13m25j06f-hnix-store-remote-0.3.0.0-doc/share/doc/hnix-store-remote-0.3.0.0 --with-gcc=gcc --package-db=/build/package.conf.d --ghc-options=-j8 +RTS -A64M -RTS --disable-split-objs --enable-library-profiling --profiling-detail=exported-functions --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --disable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --ghc-option=-split-sections --extra-lib-dirs=/nix/store/diaimgl27ri62wkf44jxan3m5vpj4f1y-ncurses-6.2/lib --extra-lib-dirs=/nix/store/zb3fx7zx5pafnqk1c6nxk1fy84k3cnxz-libffi-3.3/lib --extra-lib-dirs=/nix/store/gk72n6pafkh603q9syxgj0i3pwp4fqf4-gmp-6.2.1/lib
Using Parsec parser
Configuring hnix-store-remote-0.3.0.0...
...
building
Preprocessing library for hnix-store-remote-0.3.0.0..
Building library for hnix-store-remote-0.3.0.0..
[1 of 8] Compiling System.Nix.Store.Remote.Binary ( src/System/Nix/Store/Remote/Binary.hs, dist/build/System/Nix/Store/Remote/Binary.o, dist/build/System/Nix/Store/Remote/Binary.dyn_o )
[2 of 8] Compiling System.Nix.Store.Remote.Builders ( src/System/Nix/Store/Remote/Builders.hs, dist/build/System/Nix/Store/Remote/Builders.o, dist/build/System/Nix/Store/Remote/Builders.dyn_o )
[3 of 8] Compiling System.Nix.Store.Remote.Parsers ( src/System/Nix/Store/Remote/Parsers.hs, dist/build/System/Nix/Store/Remote/Parsers.o, dist/build/System/Nix/Store/Remote/Parsers.dyn_o )

src/System/Nix/Store/Remote/Parsers.hs:37:13: error:
    • Variable not in scope:
        decodeBase32 :: Data.Text.Text -> Either String (Digest 'SHA256)
    • Perhaps you meant ‘encodeBase32’ (imported from System.Nix.Hash)
   |
37 |   digest <- decodeBase32 @'SHA256 <$> parseHash
   |             ^^^^^^^^^^^^

src/System/Nix/Store/Remote/Parsers.hs:49:20: error:
    Variable not in scope:
      mkNamedDigest
        :: Data.Text.Text
           -> Data.Text.Text -> Either String SomeNamedDigest
   |
49 | parseTypedDigest = mkNamedDigest <$> parseHashType <*> parseHash
   |                    ^^^^^^^^^^^^^
[4 of 8] Compiling System.Nix.Store.Remote.Types ( src/System/Nix/Store/Remote/Types.hs, dist/build/System/Nix/Store/Remote/Types.o, dist/build/System/Nix/Store/Remote/Types.dyn_o )
[5 of 8] Compiling System.Nix.Store.Remote.Util ( src/System/Nix/Store/Remote/Util.hs, dist/build/System/Nix/Store/Remote/Util.o, dist/build/System/Nix/Store/Remote/Util.dyn_o )

src/System/Nix/Store/Remote/Util.hs:27:1: error:
    Could not find module ‘System.Nix.Build’
    Perhaps you meant System.Nix.Util (from hnix-store-core-0.2.0.0)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
27 | import           System.Nix.Build
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

builder for '/nix/store/fwg6v2p9f9mnrjvq6836v40pwbi69ai1-hnix-store-remote-0.3.0.0.drv' failed with exit code 1
error: build of '/nix/store/fwg6v2p9f9mnrjvq6836v40pwbi69ai1-hnix-store-remote-0.3.0.0.drv' failed

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 1 above)
Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-common.nix:831]: Ah, thanks for the explanation!

Comment (Anton-Latukha) [reply pkgs/development/haskell-modules/configuration-common.nix:831]: That is probably too opinionated and too in depth comment about a very sensitive issue from me, who was not participating in all that. Just wanted to refresh your memory on that `cryptohash-*` family and why there is a `hs-cryptohash` and `cryptohash-*` and `cryptonite` and why they use the same hierarchical module names. And that it is probable that you receive some set of PRs for projects that are variations on the theme.

**Comment by Anton-Latukha** [comment]
@cdepillabout
You are right, `0.3` does not build.

Please help to understand what is the best algorithm to do here:

I arrived because we made a `hnix-store-core` and `hnix-store-remote` both `0.4` releases to Hackage.

My friends released a broken `0.3` previously, so I went in the project and made sure to make things work further after, deployed CI both Cabal builds and Nix builds, and did stuff to made sure that the next releases would be assured to work.

Right after every Haskell project release I involved in - I do what is proper - I open Nixpkgs and check the descriptions of those projects we just released to Hackage, in Nixpkgs.
Knowing that Stackage LTS almost fully automatically updates every midnight UTC. And that everything else gets managed by you and the team through the week, and that Peti merges it all on Friday, after he made his updates and work.

So what is a proper way to do it.
Then guys released the `0.4` on Hackage.

The situation we have with `*-remote` is more complex, because guys priorly made a number of broken releases they so far not fixed.

But all that does not touches the main question and procedure. Right after we made releases - I went and cleaned up things in the Nixpkgs `haskell-updates` branch, so on Friday after Hackage regeneration and the merges the `hnix-store-remote 0.4` would work.

What do you recommend. I like to do this right after releases, and I even went a way to check it directly during development: https://github.com/marketplace/actions/automatic-haskell-project-integration-test-to-nixpkgs. But `*-store-*` is a monorepo, and so far I need to adjust the action code to support monorepos.

But overall. If person arrives with patches right after the release - we have *this problem. And doing/shipping it a 1.5 week later seems like an additional hustle to keep in mind and shedule, and with that it creates that timeframe in-between Hackage release and the Nixpkgs merge when the normal Hackage release may be still marked broken etc in Nixpkgs for a weak or so.

So far, what I made is this action - that moves the Nixpkgs store integration check before the project release, so in the upstream we always have a pulse of clean integration of Hackage release into Nixpkgs.

What we can do to ease the process we have with Hackage&`haskell-updates` misalignment.
Maybe some way to run a test that takes the Nixpkgs `master`, autoregenerates Hackage DB, and tries the patch on it.
I would think about these processes and interested in what thoughts you provide.

**Comment by Anton-Latukha** [comment]
Sorry about rambling, sometimes I write too much and I do not know how I can compress it to Shannon information boundary more.

**Comment by cdepillabout** [comment]
@Anton-Latukha Hey, sorry, I don't quite understand what needs to be done here.

Could you summarize what needs to happen for `hnix-store-remote` in order to get it building here in nixpkgs?  It looks like `hnist-store-remote-0.4` needs `hnix-store-core-0.4`, but `hnix-store-core` is locked to version 0.2 in Stackage nightly: https://github.com/commercialhaskell/stackage/issues/5766

I'd suggest trying to fix https://github.com/commercialhaskell/stackage/issues/5766 before trying to get `hnix-store-remote` working here in nixpkgs.

**Comment by Anton-Latukha** [comment]
Ok. Thanks.

We are not even aware `hnix-store-core` is in Stackage Nightly, and I have no idea why it needs to be there, Store* has no reverse-deps except us, and it is stupid to publish the Nix parser/interpreter to Stackage, especially in this still rough state. And you probably know how it is when people try to prematurely sneak the projects you develop into the Stackage, project flops in and out of Stackage and you take the heat from that oscillation.


If I knew - I'd deployed the check for it.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 2]
```diff
@@ -222,6 +222,9 @@ self: super: {
 
   });
 
+  # The project by definition uses the network for tests.
+  hnix-store-remote = dontCheck super.hnix-store-remote;
```
Comment (sorki) [inline pkgs/development/haskell-modules/configuration-common.nix:226]: Not really needed - the tests are disabled by default due to their requirements (which and nix binaries) but don't need network access. All testing is done in namespace using temporary nix store and local nix-daemon.

Comment (Anton-Latukha) [reply pkgs/development/haskell-modules/configuration-common.nix:226]: Great info. I thought the tests are effectful to the `/nix/store`.

Anyway, this PR would probably wait for a while.

Please feel free to notify me or open parallel PR, on yours decision.

**Comment by Anton-Latukha** [comment]
Used `hnix-store-core 0_4_0_0` around until Stackage unpin would happen.

`sorki` said the `hnix-store-remote` can run the test suite, so with `hnix-store-core 0_4_0_0` supplied after the expression regeneration the `hnix-store-remote` should build fine.

Current patch state addresses the current situation in projects.
